oxidation/libparsec/crates/types/Cargo.toml

```diff
@@ -48,6 +42,8 @@ hex-literal = { version = "0.4.1", optional = true }
 rstest = { version = "0.17.0", optional = true }
 
 [dev-dependencies]
+libparsec_types = { path = ".", features = ["test-fixtures"] }
```

Comment:
does it break rust anaylizer ?

Reply:
Regarding compilation time, I don't see an impact:
```
$ git checkout master
$ cargo clean -p libparsec_types
$ cargo test libparsec_types --no-run
[...]
    Finished test [unoptimized + debuginfo] target(s) in 8.87s
$ cargo clean -p libparsec_types
$ cargo build -p libparsec_types
[...]
    Finished dev [unoptimized + debuginfo] target(s) in 6.25s
$ git checkout rework-imports
$ cargo clean -p libparsec_types
$ cargo test libparsec_types --no-run
[...]
    Finished test [unoptimized + debuginfo] target(s) in 8.62s
$ cargo clean -p libparsec_types
$ cargo build -p libparsec_types
[...]
    Finished dev [unoptimized + debuginfo] target(s) in 6.37s
```

And for the full compilation (excluding 3rd party crates)
```
$ git checkout master
$ export CRATES_TO_CLEAN="-p libparsec_client_connection -p libparsec_core_fs -p libparsec_platform_device_loader -p libparsec_serialization_format -p libparsec_tests_macros -p libparsec_client_high_level_api -p libparsec_crypto -p libparsec_platform_storage -p libparsec_testbed -p libparsec_types -p libparsec_core -p libparsec_platform_async -p libparsec_protocol -p libparsec_tests_fixtures -p libparsec"
$ cargo clean $CRATES_TO_CLEAN
$ cargo build
[...]
    Finished dev [unoptimized + debuginfo] target(s) in 20.48s
$ cargo clean $CRATES_TO_CLEAN
$ cargo test --no-run
[...]
    Finished test [unoptimized + debuginfo] target(s) in 20.45s
$ git checkout rework-imports
$ cargo clean $CRATES_TO_CLEAN
$ cargo build
[...]
    Finished dev [unoptimized + debuginfo] target(s) in 19.33s
$ cargo clean $CRATES_TO_CLEAN
$ cargo test --no-run
[...]
    Finished test [unoptimized + debuginfo] target(s) in 20.35s
```

(that seems logical given I've added a dependency of `libparsec_types` on itself)

For issue with rust-analyzer, I don't see issue on my end: my IDE doesn't show errors, and I can ctrl+click on fixtures from the test file (which correctly brings me to the location of the definition of this feature within the source code)
Can you provide an example of what is broken ?

Reply:
We have an issue whose goal is to remove cyclic dependencies (#4308).
Introducing new one doesn't help, currently it seems to be fine until it won't.

Reply:
This seems to be a misunderstanding on cyclic dependencies:
1 - cyclic dependency is not a bad thing per-se
2 - cyclic dependency is allow by Rust and compile fine
3 - cyclic dependency can make rust-analyzer lost

The important thing is what we call cyclic dependencies here are not real cyclic dependencies: it is dependencies between the normal and the test flavor of crates, typically:

not allowed by Rust: `foo <- bar <- foo`
allowed by Rust but Analyzer struggles with: `foo <- bar <- foo (test)`
allowed by Rust & fine with Analyzer: `foo (with feature X) <- foo(test)`
allowed by Rust & fine with Analyzer: `foo <- foo(test)` (this is the default behavior btw)

So in the end we want to avoid the `foo <- bar <- foo (test)` kind of cyclic dependency (not because it is a bad pattern, but because of shortcoming in Rust-Analyzer) but can use the `foo (with feature X) <- foo(test)` (because Rust-Analyzer handles it fine)

Of course if in the future Rust-Analyzer better handle cyclic dependencies we will re-think this (and if my assumption on `foo (with feature X) <- foo(test)` is wrong we should rethink this now :smile: )